app/src/main/java/com/duckduckgo/app/browser/defaultbrowsing/prompts/DefaultBrowserPromptsExperimentImpl.kt

```diff
@@ -0,0 +1,341 @@
+            defaultBrowserPromptsDataStore.storeExperimentStage(newExperimentStage)
+
+            val action = experimentStageEvaluatorPluginPoint.getPlugins().first { it.targetCohort == activeCohortName }.evaluate(newExperimentStage)
+            if (action.showMessageDialog) {
+                _commands.send(OpenMessageDialog)
+            }
+            defaultBrowserPromptsDataStore.storeShowOverflowMenuItemState(action.showOverflowMenuItem)
+            defaultBrowserPromptsDataStore.storeHighlightOverflowMenuIconState(action.highlightOverflowMenu)
+        }
+    }
+
+    override fun onOverflowMenuOpened() {
+        appCoroutineScope.launch {
+            defaultBrowserPromptsDataStore.storeHighlightOverflowMenuIconState(highlight = false)
+        }
+    }
+
+    override fun onOverflowMenuItemClicked() {
+        appCoroutineScope.launch {
+            launchBestSelectionWindow()
+        }
+    }
+
+    override fun onMessageDialogShown() {
+    }
+
+    override fun onMessageDialogDismissed() {
+    }
+
+    override fun onMessageDialogConfirmationButtonClicked() {
+        appCoroutineScope.launch {
+            launchBestSelectionWindow()
+        }
+    }
+
+    override fun onMessageDialogNotNowButtonClicked() {
+    }
+
+    override fun onSystemDefaultBrowserDialogShown() {
+        browserSelectionWindowFallbackDeferred = appCoroutineScope.async {
+            delay(FALLBACK_TO_DEFAULT_APPS_SCREEN_THRESHOLD_MILLIS)
+        }
+    }
+
+    override fun onSystemDefaultBrowserDialogSuccess() {
+    }
+
+    override fun onSystemDefaultBrowserDialogCanceled() {
+        appCoroutineScope.launch {
+            if (browserSelectionWindowFallbackDeferred?.isActive == true) {
+                browserSelectionWindowFallbackDeferred?.cancel()
+                launchSystemDefaultAppsActivity()
+            }
+        }
+    }
+
+    override fun onSystemDefaultAppsActivityOpened() {
+    }
+
+    override fun onSystemDefaultAppsActivityClosed() {
+    }
+
+    private suspend fun launchBestSelectionWindow() {
+        val command = defaultRoleBrowserDialog.createIntent(applicationContext)?.let {
+            Command.OpenSystemDefaultBrowserDialog(intent = it)
+        }
+        if (command != null) {
+            _commands.send(command)
+        } else {
+            launchSystemDefaultAppsActivity()
+        }
+    }
+
+    private suspend fun launchSystemDefaultAppsActivity() {
+        val command = Command.OpenSystemDefaultAppsActivity(DefaultBrowserSystemSettings.intent())
+        _commands.send(command)
+    }
+
+    private suspend fun DefaultBrowserPromptsFeatureToggles.parseFeatureSettings(): FeatureSettings? = withContext(dispatchers.io()) {
+        defaultBrowserAdditionalPrompts202501().getSettings()?.let { settings ->
+            try {
+                featureSettingsJsonAdapter.fromJson(settings)
+            } catch (e: Exception) {
+                Timber.e(e)
+                null
```

Comment:
I’m not sure about returning null when there’s an issue parsing. How about providing a set of default config?

Reply:
Same as https://github.com/duckduckgo/Android/pull/5476#discussion_r1933600731. The experiment is fully controlled via config, so if the config doesn't work, I'd rather escape than hardcode defaults.